helper/schema/resource_data_test.go

```diff
@@ -3915,6 +3919,162 @@ func TestResourceData_nonStringValuesInMap(t *testing.T) {
+				"ConfigAttribute": cty.NumberIntVal(42),
+			}),
+			Path:  cty.GetAttrPath("invalid_root_path"),
+			Value: cty.DynamicVal,
+			ExpectedDiags: diag.Diagnostics{
+				{
+					Severity: diag.Error,
+					Summary:  "Invalid config path",
+					Detail: "The Terraform Provider unexpectedly provided a path that does not match the current schema. " +
+						"This can happen if the path does not correctly follow the schema in structure or types. " +
+						"Please report this to the provider developers. \n\n" +
+						"Cannot find config value for given path.",
+					AttributePath: cty.Path{
+						cty.GetAttrStep{Name: "invalid_root_path"},
+					},
+				},
+			},
+		},
+		"root level attribute": {
+			RawConfig: cty.ObjectVal(map[string]cty.Value{
+				"ConfigAttribute": cty.NumberIntVal(42),
+			}),
+			Path:  cty.GetAttrPath("ConfigAttribute"),
+			Value: cty.NumberIntVal(42),
+		},
+		"root level set attribute": {
+			RawConfig: cty.ObjectVal(map[string]cty.Value{
+				"ConfigAttribute": cty.SetVal([]cty.Value{
+					cty.StringVal("valueA"),
+					cty.StringVal("valueB"),
+				}),
+			}),
+			Path:  cty.GetAttrPath("ConfigAttribute").Index(cty.StringVal("valueA")),
+			Value: cty.StringVal("valueA"),
+		},
+		"root level list attribute": {
+			RawConfig: cty.ObjectVal(map[string]cty.Value{
+				"ConfigAttribute": cty.ListVal([]cty.Value{
+					cty.StringVal("valueA"),
+					cty.StringVal("valueB"),
+				}),
+			}),
+			Path:  cty.GetAttrPath("ConfigAttribute").IndexInt(0),
+			Value: cty.StringVal("valueA"),
+		},
+		"root level map attribute": {
+			RawConfig: cty.ObjectVal(map[string]cty.Value{
+				"ConfigAttribute": cty.MapVal(map[string]cty.Value{
+					"mapA": cty.StringVal("valueA"),
+					"mapB": cty.StringVal("valueB"),
+				}),
+			}),
+			Path:  cty.GetAttrPath("ConfigAttribute").IndexString("mapB"),
+			Value: cty.StringVal("valueB"),
+		},
+		"list nested block attribute - get attribute value": {
+			RawConfig: cty.ObjectVal(map[string]cty.Value{
+				"list_nested_block": cty.ListVal([]cty.Value{
+					cty.ObjectVal(map[string]cty.Value{
+						"ConfigAttribute": cty.StringVal("valueA"),
+					}),
+					cty.ObjectVal(map[string]cty.Value{
+						"ConfigAttribute": cty.StringVal("valueB"),
+					}),
+				}),
+			}),
+			Path:  cty.GetAttrPath("list_nested_block").IndexInt(1).GetAttr("ConfigAttribute"),
+			Value: cty.StringVal("valueB"),
+		},
+		"set nested block attribute - get attribute value": {
+			RawConfig: cty.ObjectVal(map[string]cty.Value{
+				"set_nested_block": cty.SetVal([]cty.Value{
+					cty.ObjectVal(map[string]cty.Value{
+						"ConfigAttribute": cty.StringVal("valueA"),
+					}),
+					cty.ObjectVal(map[string]cty.Value{
+						"ConfigAttribute": cty.StringVal("valueB"),
+					}),
+				}),
+			}),
+			Path: cty.GetAttrPath("set_nested_block").Index(cty.ObjectVal(map[string]cty.Value{
+				"ConfigAttribute": cty.StringVal("valueB"),
+			})).GetAttr("ConfigAttribute"),
```

Comment:
Feels like for sets it'd just be easier for provider developers to grab the root, but glad it works!